File: src/main/java/org/cubexmc/metro/train/TrainPhysicsController.java
package org.cubexmc.metro.train;

import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.block.data.BlockData;
import org.bukkit.block.data.Powerable;
import org.bukkit.entity.Minecart;
import org.bukkit.util.Vector;
import org.cubexmc.metro.util.LocationUtil;

/**
 * Minecart speed, launch and stall recovery calculations for train sessions.
 */
public class TrainPhysicsController {

    public void applyApproachBraking(Minecart minecart, double distance, double defaultMaxSpeed) {
        double minSpeed = 0.1;
        double speedRatio = Math.min(1.0, distance / 15.0);
        double targetSpeed = minSpeed + (defaultMaxSpeed - minSpeed) * Math.pow(speedRatio, 0.7);

        double currentMaxSpeed = minecart.getMaxSpeed();
        if (currentMaxSpeed > 0.0) {
            minecart.setMaxSpeed(Math.min(currentMaxSpeed, targetSpeed));
        }
    }

    public Vector initMinecartVelocity(Minecart minecart, float yaw) {
        Location location = minecart.getLocation();
        Block block = location.getBlock();
        BlockData blockData = block.getBlockData();

        if (block.getType() != Material.POWERED_RAIL || !(blockData instanceof Powerable powerable)
                || !powerable.isPowered()) {
            return null;
        }

        double launchRad = Math.toRadians(yaw);
        Vector direction = new Vector(
                -Math.sin(launchRad),
                0,
                Math.cos(launchRad));
        Vector normalizedDirection = direction.clone().normalize();
        minecart.setVelocity(direction.multiply(0.4));
        return normalizedDirection;
    }

    public boolean canRecoverStalledMinecart(TrainSession session) {
        Minecart minecart = session.getMinecart();
        return minecart != null
                && !minecart.isDead()
                && minecart.isValid()
                && session.getState() == TrainMovementTask.TrainState.MOVING_BETWEEN_STATIONS
                && minecart.getMaxSpeed() > 0.0
                && LocationUtil.isOnRail(minecart.getLocation())
                && session.getLastTravelDirection() != null;
    }

    public boolean isBelowCruiseSpeed(Minecart minecart, double minCruiseSpeed) {
        Vector velocity = minecart.getVelocity();
        double horizontalSpeed = Math.sqrt(velocity.getX() * velocity.getX() + velocity.getZ() * velocity.getZ());
        return horizontalSpeed < minCruiseSpeed;
    }

    public Vector buildAssistVelocity(Vector lastTravelDirection, double targetSpeed) {
        return lastTravelDirection.clone().normalize().multiply(targetSpeed);
    }

    public double resolveAssistSpeed(Minecart minecart, double configuredSpeed, double minCruiseSpeed) {
        double maxSpeed = minecart.getMaxSpeed();
        double targetSpeed = configuredSpeed > 0.0 ? Math.min(configuredSpeed, maxSpeed) : maxSpeed;
        return Math.max(minCruiseSpeed, targetSpeed);
    }
}

File: src/main/java/org/cubexmc/metro/train/TrainPhysicsController.kt
package org.cubexmc.metro.train

import kotlin.math.cos
import kotlin.math.max
import kotlin.math.min
import kotlin.math.pow
import kotlin.math.sin
import kotlin.math.sqrt
import org.bukkit.Material
import org.bukkit.block.data.Powerable
import org.bukkit.entity.Minecart
import org.bukkit.util.Vector
import org.cubexmc.metro.util.LocationUtil

/**
 * Minecart speed, launch and stall recovery calculations for train sessions.
 */
class TrainPhysicsController {
    fun applyApproachBraking(minecart: Minecart, distance: Double, defaultMaxSpeed: Double) {
        val minSpeed = 0.1
        val speedRatio = min(1.0, distance / 15.0)
        val targetSpeed = minSpeed + (defaultMaxSpeed - minSpeed) * speedRatio.pow(0.7)

        val currentMaxSpeed = minecart.maxSpeed
        if (currentMaxSpeed > 0.0) {
            minecart.maxSpeed = min(currentMaxSpeed, targetSpeed)
        }
    }

    fun initMinecartVelocity(minecart: Minecart, yaw: Float): Vector? {
        val location = minecart.location
        val block = location.block
        val blockData = block.blockData

        if (block.type != Material.POWERED_RAIL || blockData !is Powerable || !blockData.isPowered) {
            return null
        }

        val launchRad = Math.toRadians(yaw.toDouble())
        val direction = Vector(
            -sin(launchRad),
            0.0,
            cos(launchRad),
        )
        val normalizedDirection = direction.clone().normalize()
        minecart.velocity = direction.multiply(0.4)
        return normalizedDirection
    }

    fun canRecoverStalledMinecart(session: TrainSession): Boolean {
        val minecart = session.minecart
        return minecart != null &&
            !minecart.isDead &&
            minecart.isValid &&
            session.state == TrainMovementTask.TrainState.MOVING_BETWEEN_STATIONS &&
            minecart.maxSpeed > 0.0 &&
            LocationUtil.isOnRail(minecart.location) &&
            session.lastTravelDirection != null
    }

    fun isBelowCruiseSpeed(minecart: Minecart, minCruiseSpeed: Double): Boolean {
        val velocity = minecart.velocity
        val horizontalSpeed = sqrt(velocity.x * velocity.x + velocity.z * velocity.z)
        return horizontalSpeed < minCruiseSpeed
    }

    fun buildAssistVelocity(lastTravelDirection: Vector, targetSpeed: Double): Vector =
        lastTravelDirection.clone().normalize().multiply(targetSpeed)

    fun resolveAssistSpeed(minecart: Minecart, configuredSpeed: Double, minCruiseSpeed: Double): Double {
        val maxSpeed = minecart.maxSpeed
        val targetSpeed = if (configuredSpeed > 0.0) min(configuredSpeed, maxSpeed) else maxSpeed
        return max(minCruiseSpeed, targetSpeed)
    }
}
